src/Microsoft.VisualStudio.Threading/JoinableTaskFactory.cs

```diff
@@ -166,6 +166,39 @@ public MainThreadAwaitable SwitchToMainThreadAsync(CancellationToken cancellatio
             return new MainThreadAwaitable(this, this.Context.AmbientTask, cancellationToken);
         }
 
+        /// <summary>
+        /// Gets an awaitable whose continuations execute on the synchronization context that this instance was initialized with,
+        /// in such a way as to mitigate both deadlocks and reentrancy.
+        /// </summary>
+        /// <param name="alwaysYield">A value indicating whether the caller should yield even if
+        /// already executing on the main thread.</param>
+        /// <param name="cancellationToken">
+        /// A token whose cancellation will immediately schedule the continuation
+        /// on a threadpool thread.
+        /// </param>
+        /// <returns>An awaitable.</returns>
+        /// <remarks>
+        /// <example>
+        /// <code>
+        /// private async Task SomeOperationAsync() {
+        ///     // on the caller's thread.
+        ///     await DoAsync();
+        ///
+        ///     // Now switch to a threadpool thread explicitly.
+        ///     await TaskScheduler.Default;
+        ///
+        ///     // Now switch to the Main thread to talk to some STA object.
+        ///     await this.JobContext.SwitchToMainThreadAsync(alwaysYield: true);
```

Comment:
This sample isn't particularly compelling because it will always yield anyway. Can you include a  sample that demonstrates when it would be interesting to use this overload? (at least that's already on the main thread or in an ambiguous state)?

Reply:
💭 A compelling example would be great. I actually *prefer* to not use this in favor of either `alwaysYield: false` (default behavior) or `Task.Yield` (always yield, but stay on current thread). This method's primary advantage is supporting the incremental conversion of code using `SynchronizationContext.Post` or other "always yield" methods to vs-threading.

Reply:
I've updated the sample. Since I don't have permission to push to your source branch, I'll complete the PR with my commit on top locally.